stdlib/runtime/numbers.gr

```diff
@@ -0,0 +1,1336 @@
+  let x = WasmI32.fromGrain(x)
+  let result = if (isSimpleNumber(x)) {
+    newRational(untagSimple(x), 1n)
+  } else {
+    let tag = boxedNumberTag(x)
+    if (WasmI32.eq(tag, Tags._GRAIN_RATIONAL_BOXED_NUM_TAG)) {
+      x
+    } else if (WasmI32.eq(tag, Tags._GRAIN_INT32_BOXED_NUM_TAG)) {
+      newRational(boxedInt32Number(x), 1n)
+    } else if (WasmI32.eq(tag, Tags._GRAIN_INT64_BOXED_NUM_TAG)) {
+      newRational(coerceInt32(x), 1n)
+    } else {
+      throwNotRational(x)
+    }
+  }
+  WasmI32.toGrain(result): Rational
+}
+
+export let coerceNumberToFloat32 = (x: Number) => {
+  let x = WasmI32.fromGrain(x)
+  let result = if (!isSimpleNumber(x) && WasmI32.eq(boxedNumberTag(x), Tags._GRAIN_FLOAT32_BOXED_NUM_TAG)) {
+    // avoid extra malloc
+    x
+  } else {
+    newFloat32(coerceFloat32(x))
+  }
+  WasmI32.toGrain(result): Float32
+}
+
+export let coerceNumberToFloat64 = (x: Number) => {
+  let x = WasmI32.fromGrain(x)
+  let result = if (!isSimpleNumber(x) && WasmI32.eq(boxedNumberTag(x), Tags._GRAIN_FLOAT64_BOXED_NUM_TAG)) {
+    // avoid extra malloc
+    x
+  } else {
+    newFloat64(coerceFloat64(x))
+  }
+  WasmI32.toGrain(result): Float64
+}
+
+export let coerceInt32ToNumber = (x: Int32) => {
+  WasmI32.toGrain(WasmI32.fromGrain(x)): Number
+}
+
+export let coerceInt64ToNumber = (x: Int64) => {
+  WasmI32.toGrain(WasmI32.fromGrain(x)): Number
+}
+
+export let coerceRationalToNumber = (x: Rational) => {
+  WasmI32.toGrain(WasmI32.fromGrain(x)): Number
+}
+
+export let coerceFloat32ToNumber = (x: Float32) => {
+  WasmI32.toGrain(WasmI32.fromGrain(x)): Number
+}
+
+export let coerceFloat64ToNumber = (x: Float64) => {
+  WasmI32.toGrain(WasmI32.fromGrain(x)): Number
+}
+
+/// USER-EXPOSED CONVERSION FUNCTIONS
+
+export let convertExactToInexact = (x: Number) => {
+  x
+}
+
+let convertInexactToExact = (x) => {
+  if (isSimpleNumber(x)) {
+    x
+  } else {
+    let tag = boxedNumberTag(x)
+    if (WasmI32.eq(tag, Tags._GRAIN_INT32_BOXED_NUM_TAG) || WasmI32.eq(tag, Tags._GRAIN_INT64_BOXED_NUM_TAG) || WasmI32.eq(tag, Tags._GRAIN_RATIONAL_BOXED_NUM_TAG)) {
+      x
+    } else {
+      match (tag) {
+        t when WasmI32.eq(t, Tags._GRAIN_INT32_BOXED_NUM_TAG) => x,
+        t when WasmI32.eq(t, Tags._GRAIN_INT64_BOXED_NUM_TAG) => x,
+        t when WasmI32.eq(t, Tags._GRAIN_RATIONAL_BOXED_NUM_TAG) => x,
+        t when WasmI32.eq(t, Tags._GRAIN_FLOAT32_BOXED_NUM_TAG) => {
+          reducedInteger(WasmI64.truncF32S(WasmF32.nearest(boxedFloat32Number(x))))
+        },
+        t when WasmI32.eq(t, Tags._GRAIN_FLOAT64_BOXED_NUM_TAG) => {
+          reducedInteger(WasmI64.truncF64S(WasmF64.nearest(boxedFloat64Number(x))))
+        },
+        _ => {
+          fail "Unknown boxed number tag"
+        }
+      }
+    }
+  }
+}
+
+export let convertInexactToExact = (x: Number) => {
+  WasmI32.toGrain(convertInexactToExact(WasmI32.fromGrain(x))): Number
+}
+
+export let numberPlus = (x: Number, y: Number) => {
```

Comment:
Why not export these as the operators instead of named?

Reply:
I agree. I was just trying to port 1:1, and I'd like this to be changed. I can make an issue for that.

Reply:
https://github.com/grain-lang/grain/issues/546